pkg/oci/containerize.go

```diff
@@ -150,27 +152,32 @@ func newDataTarball(source, target string, ignored []string, verbose bool) error
 			}
 		}
 
-		header, err := tar.FileInfoHeader(info, info.Name())
-		if err != nil {
-			return err
+		var link string
+		if info.Mode()&os.ModeSymlink == os.ModeSymlink {
+			if link, err = os.Readlink(path); err != nil {
+				return err
+			}
+			ok, err := isWithin(source, link)
```

Comment:
I believe this would only work for absolute links (eww). For relative you need to evaluate the link relative to the link file parent dir first.

Consider:
```
source
├── a_directory
│   └── a.lnk -> ../a.txt
└── a.txt

```
The link here does not escape source but I believe this could would evaluate it as escaping.

Reply:
And this is why I dislike test on some auxiliary functions. There should be a test rather checking whole tar output.

Reply:
We need something more like:
```go
if isRelative(link) {
    ok, err := isWithin(source, join(parentOf(path), link))
} else {
    ok, err := isWithin(source, link)
}
```

Reply:
Also if we do allow absolute symliknks (eww). then you need to force relativise them.

Reply:
Thanks for the review!  I will resubmit something more compartmentalized and tested

> I believe this would only work for absolute links
The `isWithin` function does resolve the link to its absolute path, so it should work on symlinks.
However, I agree in principle and will add a higher-level test which actually uses symlinks as you suggested to make sure it works as expected.

> consider: 
> source
> ├── a_directory
> │   └── a.lnk -> ../a.txt
> └── a.txt
> The link here does not escape source but I believe this code would falsely evaluate it as escaping

This resolves to (for example if we're in `/var` ) `parent="/var/source/" child = "/var/source/a.txt".
When I tried this case with the included `isWithin` test cases, it does seem to correctly return `true`.
Again, my symlink-tar-fu is weak, so I will add a higher-level test and maybe package which uses actual links as you suggested.

> if we do allow absolute symlinks [..] relativise them

IIUC this is an error on the part of the user (path within a container will usually not match that of the host).  However, rather than alter what they are adding, I would prefer it simply not work as expected, giving them an error, with us at most issuing a warning.  I tend to prefer informing a user of incorrect input rather than trying to fix it for them.  However, I could be convinced otherwise if you feel strongly about rewriting a user's links.

As stated above, I will add a higher-level test which actually includes symlinks.  Are there no packages which handle this?  This is utter _minutia_ when all I want is a simple tarball of a directory with some sane defaults.

Reply:
>This resolves to (for example if we're in /var ) parent="/var/source/" child = "/var/source/a.txt".

Does it though @lkingland ?

I believe that `child = ../a.txt` on input of`isWithin()` , then inside the `isWithin()` it is absolutizsd relative to __CWD__ (not parent of the link as it should).
So if we are in `/var` then `child = /a.txt` if we are in `/var/source` then `child = /var/a.txt` both is something we do not want?
Also this means that if you invoke `func` CLI outside project dir using `--path` flag you get even worse results.

Reply:
You want something like
```diff
diff --git a/pkg/oci/containerize.go b/pkg/oci/containerize.go
index 0d812c42e..b7d8c3c6c 100644
--- a/pkg/oci/containerize.go
+++ b/pkg/oci/containerize.go
@@ -158,11 +158,16 @@ func newDataTarball(source, target string, ignored []string, verbose bool) error
 			if err != nil {
 				return err
 			}
-			ok, err := isWithin(source, link)
+			if filepath.IsAbs(link) {
+				return errors.New("absolute links are not permited")
+			}
+			lnkTgt := filepath.Join(filepath.Dir(path), link)
+			relLnkTgt, err := filepath.Rel(source, lnkTgt)
 			if err != nil {
-				return errors.Wrap(err, "error checking if link is outside function root")
-			} else if !ok {
-				return fmt.Errorf("links outside Function root disallowed.  Link %v references %v", path, link)
+				return fmt.Errorf("cannot evaluate relative path of link target: %w", err)
+			}
+			if strings.HasPrefix(relLnkTgt, ".."+string(filepath.Separator)) {
+				return errors.New("link escapes project's root")
 			}
 		}
 
```

Reply:
Thanks for the suggestion, I've incorporated it into an update along with expanding the test as mentioned earlier